### DISCUSSION TITLE
Make CostConfig's have their own copies of weights?

### DISCUSSION THREAD

**Discussion by YuhengZhi** [NVlabs/curobo#383] [Software Q&A]
We noticed that if multiple costs are initialized using the same cost_cfg, their weights share the same tensors under the hood. So changing one cost's weights will implicitly change the other costs' weights sharing the same torch tensors.

**Comment by balakumar-s** [comment]
Is there a reason to share the same instance of CostConfig between costs? It's a dataclass and they are expected to be created separately per cost.

**Reply by YuhengZhi** [reply]
> share the same instance of CostConfig between costs

This appeared briefly in [here](https://github.com/NVlabs/curobo/blob/a027cbcf380df07a10e7b956990f79240e951dc1/src/curobo/rollout/arm_reacher.py#L180-L193). So 1) if link_pose_cfg isn't specified, links' pose costs would be initialized by the goal_cost_cfg (though this is a deprecated feature) and 2) if multiple non-ee links have pose costs they would be sharing the same link_pose_cfg as well? 

Just to provide a bit of context, e.g., in sequential planning like the constrained_reacher example, if one wanted to change constraints for a particular non-ee link and they call `_link_pose_costs[link_name].hold_partial_pose`, this would change the run_vec_weight for all `_link_pose_costs` unintentionally.

However, I personally prefer being able to create multiple Cost instances using the same CostConfig like how is done here for the link_pose_costs rather than having to do something like Cost(CostConfig.clone()). 

Adding a few `clone()`'s in `CostConfig` and `PoseCostConfig`'s `__post_init__` solved the issue for me.
